### ISSUE TITLE
[RFC] Support of Date and Time Patterns in Index Names

### ISSUE THREAD

**Issue by jianghancn** [opensearch-project/data-prepper#755]
### What is the problem? 

OpenSearch Sink doesn’t support index names containing date and time patterns. In the existing implementation, the `index` value set in configuration will serve as an index name or an index prefix if ISM (Index State Management) rollover is enabled. There is no way to automatically create and send data to an index whose name clearly indicates current date and time. 

If Data Prepper users can add date and time patterns to the index configuration of the OpenSearch Sink Plugin which can accordingly create indices of smaller sizes over time, it would be a very helpful feature providing following benefits: 

* OpenSearch users are able to  optimize the active index for high ingest rates on high-performance hot nodes.
* OpenSearch users are able to  optimize for search performance on warm nodes.
* OpenSearch users are able to shift older, less frequently accessed data to less expensive cold nodes.
* OpenSearch users are able to delete data according to your retention policies by removing entire indices.

### What kind of business use case are you trying to solve? 

This document will address the following use case: 

* As a user, I want the OpenSearch Sink plugin to ingest logging data to index names auto-generated according to a given pattern like: \<index-prefix\>-%{date and time pattern}. 

#### Support of Time Pattern as Part of Index

Data Prepper will support a time pattern as part of index name. e.g. \<index-prefix\>-%{yyyy.MM.dd}. The time pattern is going to follow the same rule as [Java DataTimeFormatter](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html). 

This means in the OpenSearch Sink configuration, users can set index value in the format of “\<index-prefix\>-%{yyyy.MM.dd}”. For example, "application-log-%{yyyy.MM.dd}" is a valid value for the index parameter.

If rollover is enabled through Index State Management on OpenSearch, the whole index name would be something like:  your-index-prefix-2021-12-20-00001, your-index-prefix-2021-12-20-00002, and so on. 

#### Using formatted UTC time to replace date time pattern  in index names

Why UTC?

If Data Prepper generates index names based on the UTC time of the host on which Data Prepper runs, it would help data from different sources in different time zones to get into the same index on OpenSearch server. Otherwise, data from different timezones would be sent to different indexes even though the data is generated at the same UTC time. 

**Code sample for replacing time pattern in index**

In [IndexManager.java](https://github.com/opensearch-project/data-prepper/blob/main/data-prepper-plugins/opensearch/src/main/java/com/amazon/dataprepper/plugins/sink/opensearch/index/IndexManager.java), we will add a getIndexName() method which helps auto-generate index names according to current UTC time. 

```
    //We will initialize the DATE_TIME_PATTERN_FORMATTER only once when the OpenSearch Sink plugin is initialized. 
    private final static String TIME_PATTERN_REGULAR_EXPRESSION  = "%\\{.*?\\}";
    private final static String TIME_PATTERN_INTERNAL_EXTRACTOR_REGULAR_EXPRESSION  = "%\\{(.*?)\\}";
                DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DATE_TIME_PATTERN);
    Pattern DATE_TIME_PATTERN = Pattern.compile(TIME_PATTERN_INTERNAL_EXTRACTOR_REGULAR_EXPRESSION);
    DateTimeFormatter DATE_TIME_PATTERN_FORMATTER = DateTimeFormatter.ofPattern(DATE_TIME_PATTERN);
    
    //The following method is invoked when the OpenSearch Sink plugin is initialized and before each batch. 
    public final String getIndexName(){
        final String indexAlias = openSearchSinkConfiguration.getIndexConfiguration().getIndexAlias();
        Matcher matcher = pattern.matcher(indexAlias);
        if (matcher.find())
        {
            String timePattern = matcher.group(1);
            ZoneId utcZoneId = ZoneId.of(TimeZone.getTimeZone("UTC").getID());
            String formattedTimeString = DATE_TIME_PATTERN_FORMATTER.format(LocalDateTime.now().atZone(ZoneId.systemDefault()).withZoneSameInstant(utcZoneId));
            return indexAlias.replaceAll(TIME_PATTERN_REGULAR_EXPRESSION, formattedTimeString);
        
        }else{
            return indexAlias;
        }
    }
```

#### Other changes involved

* IndexTemplateName is going to be “\<index-prefix\>” +  "-index-template". so that for example, all indices with the name pattern like "application-log-%{yyyy.MM.dd.HH}" will use index template  "application-log-index-template". 

Currently index template names are generated using code: 

    indexTemplateName = 
        openSearchSinkConfiguration.getIndexConfiguration().getIndexAlias() 
        + "-index-template"; 

Since the indexAlias from configuration could have a date-time pattern, we need to remove the date time pattern from the index parameter value from configuration to form an index template name. 

Example: an index with a name pattern "application-log-%{yyyy.MM.dd}" would have a tempalte name: application-log-index-template
```
    private final static String TIME_PATTERN_REGULAR_EXPRESSION  = "%\\{.*?\\}";
    
    private final String getIndexPrefix(){
        final String indexAlias = openSearchSinkConfiguration.getIndexConfiguration().getIndexAlias();
        return indexAlias.replaceAll("-?" + TIME_PATTERN_REGULAR_EXPRESSION, "");
    }
    
    private final String getIndexTemplateName(){
        return getIndexPrefix() + "-index-template";
    }
```

* Index Template Pattern is going to be “<index-prefix>” + “\*”,  so that for example, all indices with the name pattern like "application-log-%{yyyy.MM.dd.HH}" will have the same index template pattern  "application-log\*". 

* Index policy name is going to be <index-prefix>” + "-policy", so that, for example, all indices with the name pattern like "application-log-%{yyyy.MM.dd.HH}" will reference the same policy file, "application-log-policy". 
* The [Logstash configuration converter](https://github.com/opensearch-project/data-prepper/issues/452) needs to be updated to be able to translate this date time pattern. 

Date and Time Pattern Validations

* Must not contain below characters:

```# \ / * ? " < > | , : ```

* An hour is the shortest period allowed for index rolling-over using the data time pattern. 
    *  For example, the Sink plugin won’t allow creating a new index every minute or second. 

#### Example of the OpenSearch Sink Configuration using this new feature
```
sink:
  - opensearch:
        <other parameters>
       index: "application-log-%{yyyy.MM.dd.HH}"
```
### Q&A

* Will this new date-time pattern support affect shards on OpenSearch server?
    * No. With this new feature, OpenSearch Sink is able to create indies of names containing date and time, e.g. “application-log-2021.2.23”.   To OpenSearch, it’s just an index name. OpenSearch handles index names like “application-log-2021.2.23” and “application-log-data” in the same way. 
* Can I still use OpenSearch Sink to create plain index names without any data time patterns? 
    * Yes. We don’t have to put the date time pattern into the index names. 
* Will this feature affect existing trace analytics features? 
    * No. The implementation is going to be backward compatible and won’t affect all existing features. 

#### We will choose Java time library over Joda for formatting date and time. 

According to this [doc](https://www.baeldung.com/joda-time), after Java8, Joda is not necessary.

However, [Logstash's index](https://www.elastic.co/guide/en/logstash/6.8/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-index) parameter uses Joda. If we use Joda, it would be an easier switch between Logstash and Data Prepper for customers. However, switching to Java time library is the trend in the industry. 

In summary, we would recommend Java time for converting time pattern to . I also would like to hear any thoughts you have.

### What are remaining open questions?

None

**Comment by dlvenable** [comment]
Is there a particular reason to use the percentage sign (`%{YYYY-MM.dd}`)? This may set a precedent in Data Prepper.

There may be value in avoiding `${}` since that might be useful for pipeline templates if Data Prepper were to add that.

**Comment by jianghancn** [comment]
@dlvenable 
* I am choose to use it because similar tools, [Logstash](https://www.elastic.co/guide/en/logstash/current/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-index) and [Filebeat](https://www.elastic.co/guide/en/beats/filebeat/current/elasticsearch-output.html#index-option-es) are also using it just for easier switching between tools. 
* Having "%{....}" allows us to clearly tell apart the index's prefix and the date-time suffix. 

Does this make sense?
